### ISSUE TITLE
Configuration from enclosing class not discovered for `@Nested` test class when enclosing class is not annotated with `@ContextConfiguration`

### ISSUE THREAD

**Issue by sbrannen** [spring-projects/spring-framework#31456]
## Overview

Given the following application and test classes, the `@Nested` test class fails due to `getMessage()` returning `"Hello!"` instead of `"Mocked!"` resulting from the fact that the static nested `@TestConfiguration` class is only discovered for the top-level enclosing test class.

Specifically, the `MergedContextConfiguration` for `TestConfigurationNestedTests` contains `classes = [example.Application, example.TestConfigurationNestedTests.TestConfig]`; whereas, the `MergedContextConfiguration` for `InnerTests` contains only `classes = [example.Application]`.

A cursory search for `@TestConfiguration` reveals that `SpringBootTestContextBootstrapper.containsNonTestComponent(...)` uses the `INHERITED_ANNOTATIONS` search strategy for merged annotations. Instead, it should likely need to make use of `TestContextAnnotationUtils` in order to provide proper support for `@NestedTestConfiguration` semantics (perhaps analogous to the use of `TestContextAnnotationUtils.searchEnclosingClass(...)` in `MockitoContextCustomizerFactory.parseDefinitions(...)`).

As a side note, if you uncomment `@Import(TestConfig.class)` both test classes will pass.

```java
package example;

// imports...

@SpringBootApplication
public class Application {

	public static void main(String[] args) {
		SpringApplication.run(Application.class, args);
	}

	@Service
	public static class GreetingService {
		public String getMessage() {
			return "Hello!";
		}
	}
}
```

```java
package example;

// imports...

@SpringBootTest
// @Import(TestConfig.class)
class TestConfigurationNestedTests {

	@Test
	void test(@Autowired GreetingService greetingService) {
		assertThat(greetingService.getMessage()).isEqualTo("Mocked!");
	}

	@Nested
	class InnerTests {

		@Test
		void test(@Autowired GreetingService greetingService) {
			assertThat(greetingService.getMessage()).isEqualTo("Mocked!");
		}
	}

	@TestConfiguration
	static class TestConfig {

		@MockBean
		GreetingService greetingService;

		@BeforeTestMethod
		void configureMock(BeforeTestMethodEvent event) {
			when(this.greetingService.getMessage()).thenReturn("Mocked!");
		}
	}

}
```

## Related Issues and Commits

- #18528
- 2244461778b10ab2b530aa43eda221eee706e7f5
- spring-projects/spring-boot#23929

**Comment by devikaachu** [comment]
uncommenting the `@Import(TestConfig.class)` annotation on the top-level test class will work around the issue.

Additionally, you could consider refactoring your test code to avoid the use of nested `@TestConfiguration` classes, if possible. For example, you could move the `TestConfig` class to the top-level test class and use `@BeforeEach` and `@AfterEach` methods to set up and tear down the mocked beans.

**Comment by wilkinsona** [comment]
This looks like a bug in Spring Framework to me. The difference in behavior described in [this comment](https://github.com/spring-projects/spring-framework/issues/31456) isn't a difference between Spring Boot and Spring Framework but a difference in Spring Framework's behavior with and without `@ContextConfiguration`:

```java
package com.example;

import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.test.context.junit.jupiter.SpringJUnitConfig;

import static org.assertj.core.api.Assertions.assertThat;

@SpringJUnitConfig
class SpringJUnitConfigNestedTests {

	@Test
	void test(@Autowired String foo) {
		assertThat(foo).isEqualTo("bar");
	}

	@Nested
	class InnerTests {

		@Test
		void test(@Autowired String foo) {
			assertThat(foo).isEqualTo("bar");
		}

	}

	@Configuration
	static class TestConfig {

		@Bean
		String foo() {
			return "bar";
		}

	}

}
```

```java
package com.example;

import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.test.context.junit.jupiter.SpringExtension;

import static org.assertj.core.api.Assertions.assertThat;

@ExtendWith(SpringExtension.class)
class SpringExtensionNestedTests {

	@Test
	void test(@Autowired String foo) {
		assertThat(foo).isEqualTo("bar");
	}

	@Nested
	class InnerTests {

		@Test
		void test(@Autowired String foo) {
			assertThat(foo).isEqualTo("bar");
		}

	}

	@Configuration
	static class TestConfig {

		@Bean
		String foo() {
			return "bar";
		}

	}

}
```

`SpringJUnitConfigNestedTests` passes because of the `@ContextConfiguration` meta-annotation and `SpringExtensionNestedTests` fails because of its absence. With `@ContextConfiguration` added to the original example or to `SpringBootNestedTests` they pass as well.

Without `@ContextConfiguration`, the bootstrapper calls `buildDefaultMergedContextConfiguration(testClass, cacheAwareContextLoaderDelegate)` which uses `Collections.singletonList(new ContextConfigurationAttributes(testClass))` as the config attributes list. 

With `@ContextConfiguration` the bootstrapper uses the result of `ContextLoaderUtils.resolveContextConfigurationAttributes(testClass)` as the config attributes list. In the latter case the `@Nested` test class becomes the outer class allowing the `@Configuration` or `@TestConfiguration` that it encloses to be found.

**Comment by sbrannen** [comment]
Reopening to ensure that `ContextLoaderUtils.resolveDefaultContextConfigurationAttributes()` honors bottom-up semantics.

**Comment by jhoeller** [comment]
Reopening for 7.1 since this change turned out to affect the Spring Boot and Spring Data test suites, unfortunately, which is likely to occur elsewhere as well.

**Comment by sbrannen** [comment]
Based on feedback from the Spring Boot and Spring Data teams, I assembled some tips that we may end up including in the upgrade guide and documentation. In order not to lose that information, I'm including a draft here.

----

If a static nested `@Configuration` class in a superclass or enclosing class was previously ignored due to this bug, after the bug fix it will no longer be ignored.

Thus, annotating the affected subclass or nested class with `@ContextConfiguration` allows you to take matters into your own hands and specify which classes in the hierarchy are actually intended to contribute context configuration.

In other words, with the bug fix, local `@Configuration` classes in a type or enclosing class hierarchy will always be processed when `@ContextConfiguration` is not declared anywhere in those hierarchies.

So, if you don't want such `@Configuration` classes to be processed, you can:

1. Remove the `@Configuration` declaration.
2. Apply `@ContextConfiguration` only where you actually want such classes to be processed.
3. Move static nested `@Configuration` classes which you do not want to be processed to standalone top-level classes so that they cannot be accidentally interpreted as _default_ configuration classes.

**Comment by davinkevin** [comment]
For your information, and while waiting for a fix, I've made a custom annotation to overcome the current bug. It should be easier to refactor after 😇

```kotlin
@ExtendWith(SpringExtension::class)
@Inherited
@SpringExtensionConfig(useTestClassScopedExtensionContext = true)
annotation class NestedSpringTest
```

It'd be very similar with Java code. 

Related to:

- #35880
- https://github.com/spring-projects/spring-boot/issues/48639

**Comment by sbrannen** [comment]
> Based on feedback from the Spring Boot and Spring Data teams, I assembled some tips that we may end up including in the upgrade guide and documentation. In order not to lose that information, I'm including a draft here.

FYI: this is now documented in the new [Default Context Configuration](https://docs.spring.io/spring-framework/reference/7.0-SNAPSHOT/testing/testcontext-framework/ctx-management/default-config.html) section of the reference manual.
